src/controlplane/client/authenticator.go

```diff
@@ -0,0 +1,133 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	v1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+)
+
+// Authenticator provides an interface to generate a authorized round tripper.
+type Authenticator interface {
+	Authenticate(endpoint config.Endpoint) (http.RoundTripper, error)
+}
+
+type authenticator struct {
+	logger                  log.Logger
+	secretListerByNamespace map[string]v1.SecretNamespaceLister
+	inClusterConfig         *rest.Config
+}
+
+// NewAuthenticator returns an Authenticator that supports plain, bearer token and mTLS.
+func NewAuthenticator(logger log.Logger, secretListerByNamespace map[string]v1.SecretNamespaceLister, inClusterConfig *rest.Config) Authenticator {
+	return &authenticator{
+		logger:                  logger,
+		secretListerByNamespace: secretListerByNamespace,
```

Comment:
I was thinking to move the creation of this map inside the authenticator, but i'm not sure if this kind of namespaced listers should already provide some sort of multi-namespace initialization (perhaps managed in the discovery package)

Reply:
what about modify the secret discovery to implement something like.
Perhaps just to simplify the thing just support array of namespaces at creation (not adding nor deleting)
```
type SecretListerer interface {
	Lister(namespace string) (listersv1.SecretNamespaceLister, error)
}
```
Then the authenticator accept this interface.
Probably the intitialization of this still could be handle by the `scraper` maybe moving the nasty loops to the `component` , this should unload the scraper i think.

Reply:
would be easier if this is implemented i guess https://github.com/kubernetes/kubernetes/issues/74415

Reply:
just pushed a commit with this approach , LMK WDYT